### PR TITLE
mgr: fix the race condition bw mgr pool and pod

### PR DESCRIPTION
**Pull request by parth-gr**
the devicehealth module gets started when we create the mgr pod, but the devicehealth needs to wait for the .mgr pool to get created so it can create a
device_health pool based on the .mgr pool crush rule

so wait for .mgr pool before proceeding to create mgr

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] Reviewed [AI guidelines](https://rook.io/docs/rook/latest/Contributing/ai-guidelines), if AI assisted with the PR.
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
  - Overwriting Ceph's configurations should be marked as breaking changes.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Review by BlaineEXE — COMMENTED** [review]
A user was asking about this just yesterday. Good to see we have a theory for resolution.

The user reports that they see the issue when they have no CephBlockPools defined as well. @parth-gr I think it might be good to manually test that corner case as well to see how this helps.

Thanks!

P.s., This also seems to be resolved by node reboots as a workaround. After huddle discussion, it seems likely that the mgr pod reboot is the actual thing that fixes it. I'm not sure what conclusions to draw from that, but it could be useful info

File: pkg/operator/ceph/cluster/cluster.go [hunk 1]
```diff
@@ -140,6 +142,18 @@ func (c *cluster) reconcileCephDaemons(rookImage string, cephVersion cephver.Cep
 		return errors.Wrap(err, "failed to execute post actions after all the ceph monitors started")
 	}
 
+	// Before entering mgr creation, we must create at least one pool based on the default stretch rule
+	// Wait for the .mgr pool to be created, which we expect is defined as a CephBlockPool
+	log.NamespacedInfo(c.Namespace, logger, "waiting for the builtin .mgr pool to be created")
+	pollInterval := 5 * time.Second
+	totalWaitTime := 2 * time.Minute
+	err = wait.PollUntilContextTimeout(c.ClusterInfo.Context, pollInterval, totalWaitTime, true, func(ctx context.Context) (bool, error) {
```
Comment (sp98) [inline pkg/operator/ceph/cluster/cluster.go:150]: We should be removing this check from the stretch cluster mode now. Else, we will be testing it twice in case of stretch clusters.

File: pkg/operator/ceph/cluster/cluster.go [hunk 2]
```diff
@@ -140,6 +142,18 @@ func (c *cluster) reconcileCephDaemons(rookImage string, cephVersion cephver.Cep
 		return errors.Wrap(err, "failed to execute post actions after all the ceph monitors started")
 	}
 
+	// Before entering mgr creation, we must create at least one pool based on the default stretch rule
+	// Wait for the .mgr pool to be created, which we expect is defined as a CephBlockPool
```
Comment (sp98) [inline pkg/operator/ceph/cluster/cluster.go:146]:
```suggestion
	// Wait for the .mgr pool to exist before starting the mgr. The devicehealth module starts with the mgr pod and needs the .mgr pool to create its device_health pool based on the .mgr pool's crush rule.
```

**Review by sp98 — CHANGES_REQUESTED** [review]
The CI faliure is suggesting that the fix is not working.  It worked for stretch cluster flow because the .mgr pool wait happens inside ConfigureArbiter(), which runs after the mgr and OSDs are already up. So the mgr is running when that wait executes. 

In the new code, the wait is before mgrs.Start().Looks like there is a circular dependency. You can't have the pool before the mgr, and you can't have the mgr without devicehealth racing.

If ceph somehow allows disabling the deviceHeath module then we can use an alternate approach: 

1. Disable the devicehealth module before starting the mgr (via Ceph config)
2. Start the mgr normally (no need to wait here)
3. Wait for the .mgr pool to be created (now possible since the mgr is running)
4. Re-enable devicehealth

**Comment by BlaineEXE** [comment]
> In the new code, the wait is before mgrs.Start().Looks like there is a circular dependency. You can't have the pool before the mgr, and you can't have the mgr without devicehealth racing.
> 
> If ceph somehow allows disabling the deviceHeath module then we can use an alternate approach:
> 
> 1. Disable the devicehealth module before starting the mgr (via Ceph config)
> 2. Start the mgr normally (no need to wait here)
> 3. Wait for the .mgr pool to be created (now possible since the mgr is running)
> 4. Re-enable devicehealth

This seems like a lot of logic to maintain for a workaround. Have we talked to the Ceph devs about this? From huddles this week, it seems to me like it could be resolved by a bugfix in the mgr or devicehealth module. WDYT?

**Comment by travisn** [comment]
Would this issue instead be resolved by #17502?

**Comment by parth-gr** [comment]
yes closing
